infra/base-images/base-runner/gocoverage/llvm-cov-rel/llvm-profraw-relative.go

```diff
@@ -0,0 +1,180 @@
+	h.PaddingBytesBeforeCounters = binary.LittleEndian.Uint64(data[32:40])
+	h.CountersSize = binary.LittleEndian.Uint64(data[40:48])
+	h.PaddingBytesAfterCounters = binary.LittleEndian.Uint64(data[48:56])
+	h.NamesSize = binary.LittleEndian.Uint64(data[56:64])
+	h.CountersDelta = binary.LittleEndian.Uint64(data[64:72])
+	h.NamesDelta = binary.LittleEndian.Uint64(data[72:80])
+	h.ValueKindLast = binary.LittleEndian.Uint64(data[80:88])
+
+	if h.CountersDelta != sectPrfCnts-sectPrfData {
+		// Rust linking adds an offset ? not seen in readelf
+		sectPrfData = h.CountersDelta - sectPrfCnts + sectPrfData
+		sectPrfCnts = h.CountersDelta
+	}
+	dataref := sectPrfData
+	relativizeAddress(data, 64, dataref, sectPrfCnts, sectPrfData)
+
+	offset := PROFRAW_HEADER_7_LEN + int(h.BinaryIdsSize)
+	for i := uint64(0); i < h.DataSize; i++ {
+		if len(data) < offset+profrawDataLen(h.ValueKindLast) {
+			return io.EOF, data
+		}
+		d := ProfrawData{}
+		d.NameRef = binary.LittleEndian.Uint64(data[offset : offset+8])
+		d.FuncHash = binary.LittleEndian.Uint64(data[offset+8 : offset+16])
+		d.CounterPtr = binary.LittleEndian.Uint64(data[offset+16 : offset+24])
+		d.FunctionPointer = binary.LittleEndian.Uint64(data[offset+24 : offset+32])
+		d.Values = binary.LittleEndian.Uint64(data[offset+32 : offset+40])
+		d.NumCounters = binary.LittleEndian.Uint32(data[offset+40 : offset+44])
+		d.NumValueSites = make([]uint16, h.ValueKindLast+1)
+		for j := 0; j <= int(h.ValueKindLast); j++ {
+			d.NumValueSites[j] = binary.LittleEndian.Uint16(data[offset+44+2*j : offset+46+2*j])
+		}
+
+		relativizeAddress(data, offset+16, dataref, sectPrfCnts, sectPrfData)
+		// cf CountersDelta -= sizeof(*SrcData); in __llvm_profile_merge_from_buffer
+		dataref += uint64(profrawDataLen(h.ValueKindLast))
+
+		offset += profrawDataLen(h.ValueKindLast)
+	}
+	return nil, data
+}
+
+func main() {
+	flag.Parse()
+
+	if len(flag.Args()) != 3 {
+		log.Fatalf("needs exactly three arguments : binary, profraw, output")
+	}
+
+	// first find llvm profile sections addresses in the elf
+	f, err := elf.Open(flag.Args()[0])
+	if err != nil {
+		log.Fatalf("failed to read elf: %v", err)
+	}
+	sectPrfCnts := uint64(0)
+	sectPrfData := uint64(0)
+	for i := range f.Sections {
+		if f.Sections[i].Name == "__llvm_prf_cnts" {
+			sectPrfCnts = f.Sections[i].Addr
+		} else if f.Sections[i].Name == "__llvm_prf_data" {
+			sectPrfData = f.Sections[i].Addr
+			// maybe rather sectPrfCntsEnd as f.Sections[i].Addr + f.Sections[i].Size for __llvm_prf_cnts
+		}
+	}
+	if sectPrfCnts == 0 || sectPrfData == 0 {
+		log.Fatalf("Elf has not __llvm_prf_cnts and __llvm_prf_data sections")
+	}
+
+	// process profraw file
+	data, err := ioutil.ReadFile(flag.Args()[1])
+	if err != nil {
+		log.Fatalf("failed to read file: %v", err)
+	}
+	err, data = relativizeProfraw(data, sectPrfCnts, sectPrfData)
+	if err != nil {
+		log.Fatalf("failed to process file: %v", err)
+	}
+
+	// write output file
```

Comment:
Please begin with capital and end with period.

Reply:
ok